### ISSUE TITLE
False positives on "barrel export" index files

### ISSUE THREAD

**Issue by abirmingham** [nadeesha/ts-prune#40]
**Steps to Reproduce:**

1. Create directory structure `src/stuff`
2. Create file `src/stuff/StringTypeAlias.ts` with contents `export type DogOrCat = 'dog' | 'cat';`
3. Create file `src/stuff/index.ts` with contents `export * from './StringTypeAlias';`
4. Create file `src/main.ts` with contents `import DogOrCat from 'stuff';`

**Expected Result:**
No output for `DogOrCat`.

**Actual Result:**
Output of `DogOrCat @ ./src/stuff/StringTypeAlias`.

I'm not sure if this is specific to tsconfig with absolute import paths or not, but I'm using `baseUrl` in my `tsconfig.json` so it may very well be.

**TSConfig.json:**
```
{
    "compileOnSave": false,
    "compilerOptions": {
        "baseUrl": "src/",
        "jsx": "react",
        "lib": ["es2017", "dom"],
        "moduleResolution": "node",
        "noErrorTruncation": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "allowUnreachableCode": false,
        "resolveJsonModule": true,
        "esModuleInterop": true,
        "sourceMap": true,
        "strict": true,
        "target": "es6",
        "module": "esNext",
        "typeRoots": ["./typings", "./node_modules/@types"],
        "paths": {
            "test/*": ["../test/*"],
            "stories/*": ["../stories/*"],
            "i18n/*": ["../i18n/*"]
        }
    },
    "include": [
        "src/**/*",
        "test/**/*",
        "stories/**/*",
        "i18n/**/*",
        "typings/**/*"
    ]
}
```

Thank you!

**Comment by abirmingham** [comment]
I also tried `src/stuff/index.ts` with contents `export { DogOrCat } from './StringTypeAlias';`, but still encountered the false positive output.

**Comment by abirmingham** [comment]
May be duplicate of #14 and/or #15

**Comment by abirmingham** [comment]
@nadeesha does this look like the correct solution? It seems to work for my case. If you think it looks correct, I'll open a PR. Thanks!


```diff
diff --git a/src/analyzer.ts b/src/analyzer.ts
index 2036110..881e7ba 100644
--- a/src/analyzer.ts
+++ b/src/analyzer.ts
@@ -3,7 +3,8 @@ import {
   Project,
   SourceFile,
   SourceFileReferencingNodes,
-  ts
+  ts,
+  ExportDeclaration
 } from "ts-morph";
 import { isDefinitelyUsedImport } from "./util/isDefinitelyUsedImport";
 import { getModuleSourceFile } from "./util/getModuleSourceFile";
@@ -37,8 +38,19 @@ function handleImportDeclaration(node: SourceFileReferencingNodes) {
   return referenced;
 }

+function handleExportDeclaration(node: SourceFileReferencingNodes) {
+  const referenced = [] as string[];
+
+  (node as ExportDeclaration).getNamedExports().map(n => {
+    referenced.push(n.getName());
+  });
+
+  return referenced;
+}
+
 const nodeHandlers = {
-  [ts.SyntaxKind.ImportDeclaration.toString()]: handleImportDeclaration
+  [ts.SyntaxKind.ImportDeclaration.toString()]: handleImportDeclaration,
+  [ts.SyntaxKind.ExportDeclaration.toString()]: handleExportDeclaration
 };

 function getExported(file: SourceFile) {
```

**Comment by nadeesha** [comment]
@abirmingham Thanks for looking at this 🙏 . 

Yeah, that looks to be it. A couple of things:

1. Could you write an integration test for this? Invoked via `yarn test:integration`. The files to reproduce will have to go in `integration/testproject/`

2. You could rewrite your `handleExportDeclaration` as:

```ts
function handleExportDeclaration(node: SourceFileReferencingNodes) {
  return (node as ExportDeclaration).getNamedExports().map(n => n.getName());
}
```

**Comment by ivosh** [comment]
@nadeesha @abirmingham I went ahead and created a pull request for this issue as I am interested in `ts-prune`'s correct functionality personally.

However it is only the first step to handle more complex scenario when an export from "barrel export" is not used anywhere (which should be in turn reported).

**Comment by nadeesha** [comment]
Fixed by #41. Thanks @abirmingham and @ivosh.
